dashboard/app/controllers/programming_environments_controller.rb

```diff
@@ -58,6 +58,7 @@ def update
   def show
     @programming_environment = ProgrammingEnvironment.find_by_name(params[:name])
     return render :not_found unless @programming_environment
+    return head :forbidden unless can?(:read, @programming_environment)
```

Comment:
For my education, how does this relate to `load_and_authorize_resource` at the top of this class?  Does the top-level call check any abilities?

Reply:
Oof this sent me down a rabbit hole.

The short answer to your question is (I think) no. However, according to the [CanCanCan docs](https://github.com/CanCanCommunity/cancancan/blob/develop/docs/changing_defaults.md#override-loading), I should be able to define my own method to find `@programming_environment`. Which just doesn't seem to work 🙃 Not sure if it's a version thing or what but it's not hitting the cutom logic without this explicit `can?` call.

I don't understand how the edit, update, etc permissions are working though. I know @dmcavoy has been looking in this area a lot recently so maybe she has an idea?

Reply:
CanCanCan tries to be smart and figure out the `@programming_environment` on its own but since you want it to find by name instead of id it probably isn't finding the programming environment as part of the load.  I think you might be able to fix this with `  load_and_authorize_resource find_by: :name`. If not you could do something similar to how we set up script where we `set_unit` and then `authorize_resource`

Reply:
Hmmm I tried that and it didn't seem to be respecting my custom logic in `ability.rb`. Let me try again today with fresh eyes though

Reply:
Okay, coming back to this. Writing a custom `set_programming_environment` method does work but it still won't respect the logic from ability.rb. When I look at [script](https://github.com/code-dot-org/code-dot-org/blob/599119f5d32edefee9dee177b98d3d2043407e4f/dashboard/app/controllers/scripts_controller.rb#L266) and [lesson](https://github.com/code-dot-org/code-dot-org/blob/599119f5d32edefee9dee177b98d3d2043407e4f/dashboard/app/controllers/lessons_controller.rb#L28), we do explicitly check that the page `can` be viewed by the user. So I'm inclined to leave this as-is (I'm adding the custom `set_programming_environment` method in another PR already)

Reply:
I think it's fine to leave it this way. 

The one you pointed to for script is used because we want to render a custom page when you can't access so its a little different case.

Reply:
Sounds good, thanks for taking a look!

Reply:
Ahh makes sense about script (and why it's also like that for courses). Thanks for the help Dani!